source/tutorial/autocomplete-component.md

```diff
@@ -0,0 +1,164 @@
+  <li {{action 'choose' item.city}}>{{item.city}}</li>
+{{/each}}
+</ul>
+```
+It contains an `input` component, that renders as a text field that the user can type in to look for properties in a given city. The `value` property of the `input` will be bound to the `filter` property in our backing JavaScript object. The `key-up` property will be bound to a `keyUp` action in our backing object, and passes the `filter` property as a parameter.
+
+It also contains a button, whose `action` parameter is bound to the `click` action in our component.
+
+Lastly, it contains an unordered list, that uses the `filteredList` property for data, and displays the `city` property of each item in the list. Clicking the list item will fire the `choose` action, which will populate the `input` field with the name of the `city` in the clicked list item.
+
+Here is what the component's JavaScript looks like:
+
+```app\components\filter-listing.js
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+  filter: null,
+  filteredList: null,
+  actions: {
+    keyUp() {
+      this.attrs.autoComplete(this.filter);
+    },
+    click() {
+      this.attrs.search(this.filter);
+    },
+    choose(city) {
+      this.set('filter',city);
+    }
+  }
+});
+
+```
+There's a property for each of the `filter` and `filteredList`, and actions as described above. What's interesting is that only the `choose` action is defined by the component. The actual logic of each of the `keyUp` and `click` actions are pulled from the component's `attrs` object, which means that those actions need to be passed in by the calling object, a pattern known as _closure actions_.
+
+To see how this works, change your `index.hbs` template to look like this:
+
+```app\templates\index.hbs
+<h1>Welcome to Super Rentals</h1>
+
+We hope you find exactly what you're looking for in a place to stay.
+<br /><br />
+{{filter-listing filteredList=filteredList 
+autoComplete=(action 'autoComplete') search=(action 'search')}}
+{{#each model as |rental-unit|}}
+  {{rental-listing rental=rental-unit}}
+{{/each}}
+
+{{#link-to 'about'}}About{{/link-to}}
+{{#link-to 'contact'}}Click here to contact us.{{/link-to}}
+```
+We've added the `filter-listing` component to our `index.hbs` template. We then pass in the functions and properties we want the `filter-listing` component to use, so that the `index` page can define some of how it wants the component to behave, and so the component can have some narrowly-defined access to aspects of the calling scope.
+
+For this to work, we need to introduce a `controller` into our app. Generate a controller for the `index` page by running the following:
+
+```shell
+ember g controller index
+```
+
+Now, define your new controller like so:
+
+```app\controllers\index.js
+import Ember from 'ember';
+
+export default Ember.Controller.extend({
+  filteredList: null,
+  actions: {
+    autoComplete(param) {
+      if(param !== undefined && param !== "") {
```

Comment:
Yeah, because otherwise if you don't check for empty/undefined, you get weird behavior when you clear out the text box.